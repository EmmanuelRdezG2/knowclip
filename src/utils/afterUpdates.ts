import { defer, Observable } from 'rxjs'
import { mergeAll } from 'rxjs/operators'

export const afterUpdates = <T>(callback: () => Promise<Observable<T>>) =>
  defer(async () => {
    await null
    return await callback()
  }).pipe(mergeAll())
