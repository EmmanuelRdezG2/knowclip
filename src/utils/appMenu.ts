import electron from 'electron'

export function setAppMenuProjectSubmenuPermissions(projectOpened: boolean) {
  const menu = electron.remote.Menu.getApplicationMenu()
  const submenu = menu && menu.getMenuItemById('File').submenu
  if (!submenu) return

  submenu.getMenuItemById('Save project').enabled = projectOpened
  submenu.getMenuItemById('Close project').enabled = projectOpened

  submenu.getMenuItemById('Open project').enabled = !projectOpened
}
