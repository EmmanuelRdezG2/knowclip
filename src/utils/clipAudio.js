import ffmpeg, { toTimestamp } from '../utils/ffmpeg'

const clipAudio = (sourceFilePath, startTime, endTime, outputFilename) => {
  return new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // later, do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on(
        'end',
        //listener must be a function, so to return the callback wrapping it inside a function
        function() {
          res(outputFilename)
        }
      )
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })
}
export default clipAudio
