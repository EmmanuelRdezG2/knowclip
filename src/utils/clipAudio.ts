import ffmpeg, { toTimestamp } from '../utils/ffmpeg'
import { existsSync } from 'fs-extra'

import { Writable } from 'stream'

export const clipAudioStream = (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): { done: Promise<string>; stream: Writable } => {
  const mp3 = ffmpeg(sourceFilePath)
    // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
    .seekInput(toTimestamp(startTime))
    .inputOptions('-to ' + toTimestamp(endTime))
    .outputOptions('-vn')
    .outputFormat('mp3')
  return {
    done: new Promise((res, rej) => {
      mp3
        .on('end', () => {
          console.log('done!', outputFilename)
          res(outputFilename)
        })
        .on('error', err => {
          console.error(err)
          rej(err)
        })
    }),
    stream: mp3.pipe(),
  }
}

const clipAudio = async (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): Promise<string> => {
  const result = await new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on('end', () => {
        res(outputFilename)
      })
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })

  if (!result || !existsSync(outputFilename)) {
    console.log({ outputFilename, sourceFilePath, startTime, endTime })
    throw new Error(`Problem clipping audio from ${sourceFilePath}`)
  }

  return outputFilename
}
export default clipAudio
