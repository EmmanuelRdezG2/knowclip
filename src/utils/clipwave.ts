export const CLIPWAVE_ID = 'waveform' as const
