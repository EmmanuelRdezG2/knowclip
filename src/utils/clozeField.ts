type RawClozeRange = {
  startIndex: number
  fullMatchText: string
  tagContents: string
}

export const unescapeClozeFields = <F extends FlashcardFields>(
  fieldsFromJson?: Partial<F>
): { fields: Partial<F>; cloze: ClozeDeletion[] } => {
  console.log('UNESCAPING!!')
  const fields = {} as F
  if (!fieldsFromJson) return { fields, cloze: [] }
  const cloze: ClozeDeletion[] = []
  for (const fn in fieldsFromJson) {
    const fieldName: keyof F = fn
    const value: string = fieldsFromJson[fieldName] as any
    if (fieldName === 'transcription') {
      const { text, clozeDeletions } = unescapeClozeField(value)
      fields[fieldName] = (text as unknown) as F[keyof F]
      cloze.push(...clozeDeletions)
    } else {
      fields[fieldName] = (value as unknown) as F[keyof F]
    }
  }

  return { fields, cloze }
}

const clozeTagsRegex = new RegExp(`<c(10|[1-9])>(.*?)</c>`, 'g')

export function unescapeClozeField(text: string) {
  const matches = [...text.matchAll(clozeTagsRegex)]

  let result = ''
  const rawRangeGroups = matches
    .reduce(
      (deletions, match) => {
        const [fullMatchText, clozeNumberRaw, tagContents] = match
        const index = typeof match.index === 'number' ? match.index : -1
        const clozeNumber = +clozeNumberRaw

        deletions[clozeNumber - 1] = deletions[clozeNumber - 1] || []
        deletions[clozeNumber - 1].push({
          startIndex: index,
          fullMatchText,
          tagContents,
        })
        return deletions
      },
      [] as RawClozeRange[][]
    )
    .filter(r => r && r.length)

  if (!rawRangeGroups.length) return { text, clozeDeletions: [] }

  const deletions: ClozeDeletion[] = rawRangeGroups.map(
    (rawRanges, rangeGroupIndex) => {
      const ranges: ClozeRange[] = []

      const firstRange = rawRanges[0]
      if (rangeGroupIndex === 0 && firstRange.startIndex > 0) {
        result += text.slice(0, firstRange.startIndex)
      }

      let i = 0
      for (const rawRange of rawRanges) {
        ranges.push({
          start: result.length,
          end: result.length + rawRange.tagContents.length,
        })
        result += rawRange.tagContents

        const nextRange = rawRanges[i + 1]
        const subsequentGapEnd = nextRange ? nextRange.startIndex : text.length
        result += text.slice(
          rawRange.startIndex + rawRange.fullMatchText.length,
          subsequentGapEnd
        )

        i++
      }

      return { ranges }
    }
  )

  return { text: result, clozeDeletions: deletions }
}
