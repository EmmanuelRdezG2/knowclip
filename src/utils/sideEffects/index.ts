import sideEffects from './module'
import spectronMocks from '../../test/spectronMocks'

const {
  mocked,
  mockFunctions: mockSideEffects,
  logMocks: logSideEffectsMocks,
} = spectronMocks('side-effect', sideEffects)

export { mockSideEffects, logSideEffectsMocks }

const { uuid, nowUtcTimestamp } =
  process.env.REACT_APP_CHROMEDRIVER || process.env.NODE_ENV === 'test'
    ? mocked
    : sideEffects

export { uuid, nowUtcTimestamp }
