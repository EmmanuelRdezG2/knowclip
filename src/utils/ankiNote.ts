import { join, basename } from 'path'
import { promises } from 'fs'
import clipAudio from '../utils/clipAudio'
import { getVideoStill } from '../utils/getVideoStill'
const { readFile } = promises

export type AnkiNoteMedia = {
  soundData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  }
  imageData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  } | null
}

export function getClozeFields(fields: string[], clozeDeletions: string) {
  return [clozeDeletions, ...fields.slice(1)]
}

export async function processNoteMedia(
  clipSpecs: ClipSpecs,
  directory: string
): AsyncResult<AnkiNoteMedia> {
  const {
    outputFilename,
    sourceFilePath,
    startTime,
    endTime,
    flashcardSpecs: { image },
  } = clipSpecs

  const clipOutputFilePath = join(directory, outputFilename)
  const clipAudioResult = await clipAudio(
    sourceFilePath,
    startTime,
    endTime,
    clipOutputFilePath
  )
  if (clipAudioResult.errors)
    return {
      errors: [
        `Could not make clip from ${sourceFilePath}`,
        ...clipAudioResult.errors,
      ],
    }
  const soundData = {
    data: () => readFile(clipOutputFilePath),
    fileName: outputFilename,
    filePath: clipOutputFilePath,
  }

  const imageResult = image
    ? await getVideoStill(image.id, sourceFilePath, image.seconds)
    : null
  if (imageResult && imageResult.errors)
    return {
      errors: [
        `Could not make still from ${sourceFilePath}`,
        ...imageResult.errors,
      ],
    }

  const imageData = imageResult
    ? {
        data: () => readFile(imageResult.value),
        fileName: basename(imageResult.value),
        filePath: imageResult.value,
      }
    : null

  return {
    value: {
      soundData,
      imageData,
    },
  }
}
