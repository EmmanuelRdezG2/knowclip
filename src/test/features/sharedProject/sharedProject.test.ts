import { Application } from 'spectron'
import { startApp, stopApp, TestSetup } from '../../spectronApp'
import openSharedProject from './openSharedProject'
import navigateBetweenMedia from './navigateBetweenMedia'
import makeFlashcardsWithSubtitles from './makeFlashcardsWithSubtitles'
import manuallyLocateAsset from './manuallyLocateAsset'
import reviewWithMissingMedia from './reviewWithMissingMedia'
import exportWithMissingMedia from './exportWithMissingMedia'
import saveAndCloseProject from './saveAndCloseProject'
import { mockSideEffects } from '../../../utils/sideEffects'

jest.setTimeout(60000)

const testId = 'sharedProject'

describe('opening a shared project', () => {
  let context: { app: Application | null; testId: string } = {
    app: null,
    testId,
  }
  let setup: TestSetup

  beforeAll(async () => {
    setup = await startApp(context)

    await mockSideEffects(setup.app, sideEffectsMocks)
  })

  test('open a shared project and locates media in local filesystem', () =>
    openSharedProject(setup))
  test('navigate between as-of-yet unloaded media', () =>
    navigateBetweenMedia(setup))
  test('make some flashcards', () => makeFlashcardsWithSubtitles(setup))
  test('manually locate missing assets', () => manuallyLocateAsset(setup))
  test('review with missing media', () => reviewWithMissingMedia(setup))
  test('export deck with missing media', () => exportWithMissingMedia(setup))
  test('save and close project', () => saveAndCloseProject(setup))

  afterAll(async () => {
    await stopApp(context)
  })
})

const sideEffectsMocks = {
  uuid: [
    '64bc9fe8-822a-4ecf-83d9-d6997029db7d',
    'b9ba2184-cb5c-4d50-98c2-568bf8e75854',
  ],
  nowUtcTimestamp: [
    '2020-01-24T15:09:02Z',
    '2020-01-24T15:09:03Z',
    '2020-01-24T15:09:18Z',
    '2020-02-02T15:26:22Z',
    '2020-02-02T15:26:22Z',
    '2020-02-02T15:26:24Z',
    '2020-02-02T15:26:24Z',
    '2020-02-02T15:26:24Z',
    '2020-02-02T15:26:26Z',
    '2020-02-02T15:26:28Z',
    '2020-02-02T15:26:29Z',
    '2020-02-02T15:26:29Z',
    '2020-02-02T15:26:30Z',
    '2020-02-02T15:26:34Z',
    '2020-02-02T15:26:34Z',
    '2020-02-02T15:26:39Z',
    '2020-02-02T15:26:39Z',
    '2020-02-02T15:26:39Z',
    '2020-02-02T15:26:39Z',
    '2020-02-02T15:26:39Z',
    '2020-02-02T15:26:39Z',
  ],
}
