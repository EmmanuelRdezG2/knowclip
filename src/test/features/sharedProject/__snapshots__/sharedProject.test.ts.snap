// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opening a shared project save and close project 1`] = `
Object {
  "clips": Array [
    Object {
      "end": 888,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Du hast eben gesagt, murrte Piggeldy, eine Katze sieht aus wie ein Schwein, stimmt's?",
        },
        "id": "2f957585-b782-45fa-b0e7-ad6b0486d9ca",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "2f957585-b782-45fa-b0e7-ad6b0486d9ca",
      "start": 603,
    },
    Object {
      "end": 1545,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Sie hat nur fast alles, was ein Schwein auch hat. Aber sie spricht anders, sie sagt »Miau, miau, miau!«",
        },
        "id": "298d069b-1b30-420c-b3b4-2728569a6718",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "298d069b-1b30-420c-b3b4-2728569a6718",
      "start": 1074,
    },
    Object {
      "end": 2048,
      "fileId": "fa462524-bfeb-4434-a506-07858cc97151",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "Schön! seufzte Piggeldy verzückt.

Schön, wie du eben »Miau!« gemacht hast. Ein Glück, dass du keine Mäuse frisst.",
        },
        "id": "505afa2a-928d-4ce1-a741-8827c4e9a864",
        "tags": Array [
          "piggeldy_cat.mp4",
        ],
        "type": "Transliteration",
      },
      "id": "505afa2a-928d-4ce1-a741-8827c4e9a864",
      "start": 1571,
    },
    Object {
      "end": 625,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Relaxing while eating bamboo grass is the best",
          "notes": "",
          "pronunciation": "sasa-o tabe-nágara nonbíri-suru-no-wa saikoo-da-naa",
          "transcription": "笹を食べながらのんびりするのは最高だなぁ",
        },
        "id": "ae3239c8-90bd-4488-95e3-bde059632348",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "ae3239c8-90bd-4488-95e3-bde059632348",
      "start": 402,
    },
    Object {
      "end": 888,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "This kid, lazing about again so early",
          "notes": "\\"Goro-goro\\" is the sound of something big rolling around.",
          "pronunciation": "mata kono ko-wa hirumá-kara góro-goro shite",
          "transcription": "またこの子は昼間からゴロゴロして",
        },
        "id": "7c348315-8f78-42fd-9be7-b05bc0aade9d",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "7c348315-8f78-42fd-9be7-b05bc0aade9d",
      "start": 756,
    },
    Object {
      "end": 2774,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "Don't try to suck me up!",
          "notes": "",
          "pronunciation": "",
          "transcription": "ああー  吸わないで",
        },
        "id": "b9ba2184-cb5c-4d50-98c2-568bf8e75854",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "b9ba2184-cb5c-4d50-98c2-568bf8e75854",
      "start": 2522,
    },
    Object {
      "end": 4058,
      "fileId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "flashcard": Object {
        "fields": Object {
          "meaning": "",
          "notes": "",
          "pronunciation": "",
          "transcription": "",
        },
        "id": "e62c823d-93c2-4a5e-8181-37e3d540c3e3",
        "tags": Array [
          "pbc",
        ],
        "type": "Transliteration",
      },
      "id": "e62c823d-93c2-4a5e-8181-37e3d540c3e3",
      "start": 3760,
    },
  ],
  "id": "e36ca6de-4893-49f5-bca1-4410ace25d46",
  "lastOpened": "2020-01-24T15:09:03Z",
  "mediaFiles": Array [
    Object {
      "durationSeconds": 58.24,
      "flashcardFieldsToSubtitlesTracks": Object {},
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "fa462524-bfeb-4434-a506-07858cc97151",
      "isVideo": true,
      "width": 480,
      "height":368,
      "name": "piggeldy_cat.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [],
      "subtitlesTracksStreamIndexes": Array [],
      "type": "MediaFile",
    },
    Object {
      "durationSeconds": 87.062,
      "flashcardFieldsToSubtitlesTracks": Object {
        "meaning": "37bd5e91-89c5-491b-9308-533c7be7338a",
        "transcription": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
      },
      "format": "mov,mp4,m4a,3gp,3g2,mj2",
      "id": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "isVideo": true,
      "width": 1920,
      "height": 1080,
      "name": "polar_bear_cafe.mp4",
      "parentId": "e36ca6de-4893-49f5-bca1-4410ace25d46",
      "subtitles": Array [
        Object {
          "id": "37bd5e91-89c5-491b-9308-533c7be7338a",
          "streamIndex": 2,
          "type": "EmbeddedSubtitlesTrack",
        },
        Object {
          "id": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
          "type": "ExternalSubtitlesTrack",
        },
      ],
      "subtitlesTracksStreamIndexes": Array [
        2,
      ],
      "type": "MediaFile",
    },
  ],
  "name": "My friend's shared project",
  "noteType": "Transliteration",
  "subtitles": Array [
    Object {
      "id": "2f7a98c0-1bf8-44cf-9b0b-4af64fb86439",
      "name": "pbc_jp.ass",
      "parentId": "535935a3-7d72-4238-87ab-1b7a413c1f71",
      "type": "ExternalSubtitlesFile",
    },
  ],
  "tags": Array [
    "piggeldy_cat.mp4",
    "pbc",
  ],
  "timestamp": "2020-01-24T15:09:18Z",
  "version": "4.1.0",
}
`;
